onnxruntime/core/graph/graph_utils.cc

```diff
@@ -191,51 +193,121 @@ static void UpdateImplicitInputNameInSubgraph(Node& node,
+                              GetNodeOutputName(node, (*it).GetSrcArgIndex()));
   }
 
-  // Get input/output edges, nodes, and node args.
-  const Node::EdgeEnd& input_edge = *node.InputEdgesBegin();
-  const Node& input_edge_node = input_edge.GetNode();
-  const NodeIndex input_edge_node_index = input_edge_node.Index();
-  const int input_edge_dst_arg = input_edge.GetSrcArgIndex();
+  // Remove output edges
+  for (const auto& edge_to_remove : output_edges) {
+    graph.RemoveEdge(edge_to_remove.src_node,
+                     edge_to_remove.dst_node,
+                     edge_to_remove.src_arg_index,
+                     edge_to_remove.dst_arg_index);
+  }
 
-  const Node::EdgeEnd& output_edge = *node.OutputEdgesBegin();
-  const Node& output_edge_node = output_edge.GetNode();
-  const NodeIndex output_edge_node_index = output_edge_node.Index();
-  const int output_edge_dst_arg = output_edge.GetDstArgIndex();
+  return output_edges;
+}
 
-  // check if the output from the node to remove is implicit input to the downstream node.
-  // if so, it's used in subgraph in that node and we have to check if it's valid to update that subgraph
-  if (OutputEdgeProvidesImplicitInput(output_edge)) {
-    // the output we need to wire to the downstream node, and the output name from the node we want to remove
-    const auto& output_name = GetNodeOutputName(input_edge_node, output_edge.GetSrcArgIndex());
-    const auto& removed_output_name = GetNodeOutputName(node, 0);
+/** Remove a node with a single incoming node. */
+static bool RemoveNodeWithSingleNodeIn(Graph& graph, Node& node) {
+  // Store info for and remove the output edges.
+  std::vector<GraphEdge> output_edges = RemoveAndReturnOutputEdges(graph, node);
 
-    Node& mutable_output_edge_node = *graph.GetNode(output_edge_node_index);
+  // Store info for the input edge.
+  const Node::EdgeEnd& input_edge_end = *node.InputEdgesBegin();
+  const auto& input_edge_node = input_edge_end.GetNode();
+  const GraphEdge input_edge{input_edge_node.Index(),
+                             node.Index(),
+                             input_edge_end.GetSrcArgIndex(),
+                             input_edge_end.GetDstArgIndex(),
+                             GetNodeOutputName(input_edge_node, input_edge_end.GetSrcArgIndex())};
 
-    if (CanUpdateImplicitInputNameInSubgraph(mutable_output_edge_node, removed_output_name, output_name)) {
-      UpdateImplicitInputNameInSubgraph(mutable_output_edge_node, removed_output_name, output_name);
-    } else {
-      // we can't safely remove this node
+  // Remove node (this will remove the input edge too).
+  graph.RemoveNode(node.Index());
+
+  // Create connections between the incoming node and the outgoing nodes of the node that we removed.
+  for (const auto& output_edge : output_edges) {
+    // Take care of subgraph inputs.
+    if (!CheckAndUpdateImplicitInputInSubgraph(graph, input_edge.arg_name, output_edge)) {
       return false;
     }
```

Comment:
Is it valid to remove the node and later discover we can't update the subgraph?

Reply:
I think it is not, good point. I should first perform the check and later the action.

Reply:
Done with latest commit.

Reply:
Don't you need to check before you remove output edges?

Reply:
It's better that I do, yes (not the end of the world since Resolve would fix the graph, but still better to not get there).